### PR TITLE
Fixing issue with default_value for custom traits

### PR DESCRIPTION
**Pull request by ssunkara1**
@rmenegaux @SylvainCorlay

### CODE REVIEW

**Comment by ssunkara1** [comment]
This addresses issue #290

File: bqplot/traits.py [hunk 1]
```diff
@@ -145,6 +145,8 @@ def _set_value(self, value):
     def set(self, obj, value):
         #TODO: We shouldnt have to overload the set because
         # numpy doesn't support == comparions
+        #TODO: This should also be removed once _buffered_state_diff is
+        # fixed in ipywidgets
```
Comment (SylvainCorlay) [inline bqplot/traits.py]: I  don't agree with this comment. The fact that NdArray emits an event on any assignment (even when the new value is equal to the old one) is a bug regardless of the _buffered_state_diff issue in ipywidgets.

This issue only exposed the problem with nd array.

File: bqplot/traits.py [hunk 2]
```diff
@@ -186,12 +188,23 @@ def _from_json(value, obj=None):
...
+    # default constructor with args as the first argument and kwargs
+    # as the second argument. The constructor of np.ndarray
+    # requires shape, data type and stride to be passed.
+    def make_dynamic_default(self):
+        return self._default_value
+
     def __init__(self, *args, **kwargs):
         self.squeeze = kwargs.pop('squeeze', True)
+        default_value = kwargs.pop('default_value', None)
+        self._default_value = self._cast(default_value)
```
Comment (SylvainCorlay) [inline bqplot/traits.py]: why not only doing
```
if 'default_value' in kwargs:
    kwargs['default_value'] = self._cast(default_value)
```
and in `make_dynamic_default`
```
return self.default_value
```

File: bqplot/traits.py (hunk 1 above)
Comment (ssunkara1) [inline bqplot/traits.py]: Well, previously, `_buffered_state_diff` only returned what trait changed. Now for the first two calls it returns everything. 

The behavior of `_buffered_state_diff` should be consistent with every call I feel. It should either return everything always or return only the changed things but not change depending on if it is the first call or the fourth.

Comment (SylvainCorlay) [inline bqplot/traits.py]: I agree that there is an issue with `_buffered_state_diff`, but the fact that the numpy array emits an event on any assignment is a bug regardless.

Therefore, even when we have fixed the `_buffered_state_diff` thing, we should keep this.

**Comment by ssunkara1** [comment]
@SylvainCorlay Yeah, I agree it is better to have the `set` over-ridden even when `_buffered_state_diff` changes. I removed that comment.  Made the other change with `default_value` as well.

**Comment by ssunkara1** [comment]
Do not merge as of now. Still testing all edge cases. I will let you know when I am good.

Thanks.

**Comment by ssunkara1** [comment]
@SylvainCorlay Ready to be reviewed.

Thanks.

**Pull request merged**